pkg/migration/azure_crs.go

```diff
@@ -0,0 +1,392 @@
+	kubeadm "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha3"
+	capiexp "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
+	ctrl "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	EncryptionSecret = "EncryptionSecret"
+)
+
+func (m *azureMigrator) createEncryptionConfigSecret(ctx context.Context) error {
+	var origEncryptionSecret *corev1.Secret
+	{
+		obj, exists := m.crs[EncryptionSecret]
+		if !exists {
+			return microerror.Mask(fmt.Errorf("encryption secret not found"))
+		}
+
+		origEncryptionSecret, ok := obj.(*corev1.Secret)
+		if !ok {
+			return microerror.Mask(fmt.Errorf("can't convert obj (%T) to %T", obj, origEncryptionSecret))
+		}
+	}
+
+	encryptionConfigTmpl := `
+kind: EncryptionConfiguration
+apiVersion: apiserver.config.k8s.io/v1
+resources:
+  - resources:
+    - secrets
+    providers:
+    - aescbc:
+        keys:
+        - name: key1
+          secret: %s
+    - identity: {}`
+
+	renderedConfig := fmt.Sprintf(encryptionConfigTmpl, origEncryptionSecret.Data["encryption"])
+
+	s := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-k8s-encryption-config", m.clusterID),
+			Namespace: "default",
+		},
+		Type: corev1.SecretTypeOpaque,
+		StringData: map[string]string{
+			"encryption": renderedConfig,
+		},
+	}
+
+	err := m.mcCtrlClient.Create(ctx, s)
+	if apierrors.IsAlreadyExists(err) {
+		// It's fine. No worries.
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (m *azureMigrator) createProxyConfigSecret(ctx context.Context) error {
+	proxyConfig := `
+apiVersion: kubeproxy.config.k8s.io/v1alpha1
+clientConnection:
+  kubeconfig: /etc/kubernetes/config/proxy-kubeconfig.yaml
+kind: KubeProxyConfiguration
+mode: iptables
+metricsBindAddress: 0.0.0.0:10249`
+
+	s := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-proxy-config", m.clusterID),
+			Namespace: "default",
+		},
+		Type: corev1.SecretTypeOpaque,
+		StringData: map[string]string{
+			"proxy": proxyConfig,
+		},
+	}
+	err := m.mcCtrlClient.Create(ctx, s)
+	if apierrors.IsAlreadyExists(err) {
+		// It's fine. No worries.
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (m *azureMigrator) createKubeadmControlPlane(ctx context.Context) error {
+	tmpl, err := template.ParseFS(templatesFS, "templates/kubeadm_controlplane_azure.yaml.tmpl")
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	cfg := struct {
+		ClusterID              string
+		ClusterCIDR            string
+		ClusterMasterIP        string
+		EtcdVersion            string
+		K8sVersion             string
+		InstallationBaseDomain string
+	}{
+		ClusterID:              m.clusterID,
+		ClusterCIDR:            "10.2.0.0/16",
+		ClusterMasterIP:        "10.2.0.4",
+		EtcdVersion:            "v3.4.13",
+		K8sVersion:             "v1.19.9",
+		InstallationBaseDomain: "gremlin.germanywestcentral.azure.gigantic.io",
+	}
```

Comment:
Here you could use a map[string]interface{} to make a bit more concise. We don't gain type safety anyway.